pkg/storage/google_cloud_storage.go

```diff
@@ -154,16 +154,16 @@ func (s *GoogleCloudStorage) UploadObject(ctx context.Context, bucket, name stri
 }
 
 // DownloadObject downloads an object from a Google Cloud Storage bucket. The caller must call Close on the returned Reader when done reading.
-func (s *GoogleCloudStorage) DownloadObject(ctx context.Context, bucket, name string) (io.ReadCloser, error) {
+func (s *GoogleCloudStorage) DownloadObject(ctx context.Context, bucket, name string) (io.ReadCloser, context.CancelFunc, error) {
 	o, ctx, cancel := s.objectHandleWithRetries(ctx, bucket, name)
-	defer cancel()
```

Comment:
This is a little bit weird, because now we're leaking our internal abstraction outwards.

The user is providing a context in the call signature, and it's up to them to manage that context (that's the rule of contexts). We are creating our own internal context for the purposes of forcing a timeout on the retries.

We want to ensure the retryContext is cancelled when the user is done reading, since that frees up the context's resources. Therefore I think we want this:

```go
type readCloserCanceller struct {
  rc io.ReadCloser
  cancelFunc context.CancelFunc
}

func (r *readCloserCanceller) Close() error {
  defer r.cancelFunc()
  return r.Close()
}

func DownloadObject() {
  r, err := o.NewReader(ctx)
  if err != nil {
    cancel()
    return nil, fmt.Errorf("failed to get google cloud storage reader: %w", err)
  }

  return &readCloserCanceller{
    rc: r,
    cancelFunc: cancel,
  }, nil
}
```

Reply:
https://github.com/abcxyz/guardian/pull/138

Reply:
Ack. The other thing I considered was fully encapsulating by accepting an argument that is a marshalling func like you did in the original impl for the json unmarshalling of tf state. Either would be better. I'll make an issue to address this. https://github.com/abcxyz/guardian/issues/137